.github/workflows/array_api.yml

```diff
@@ -105,3 +105,4 @@ jobs:
         python dev.py --no-build test -b all -t scipy.stats.tests.test_variation -- --durations 3 --timeout=60
         python dev.py --no-build test -b all -t scipy.stats.tests.test_resampling -- --durations 3 --timeout=60
         python dev.py --no-build test -b all -t scipy.optimize.tests.test_chandrupatla -- --durations 3 --timeout=60
+        python dev.py --no-build test -b jax.numpy -t scipy/signal/tests/test_signaltools.py::TestConvolve -- --durations 3 --timeout=60
```

Comment:
why only `jax.numpy`?

Reply:
We do not run cupy on CI, do we?

Reply:
No, but it's generally good to check that `array_api_strict` and `pytorch` pass the tests too, even if there is no special handling for them.

Reply:
We don't, since there's no discrete GPU for it with the default-provided runners of course. There may be a route to using MPS for Apple "GPU" per https://github.com/cirruslabs/tart/issues/501 eventually, but that doesn't help for CuPy anyway.

Perhaps Lucas is suggesting flushing through all the host-based backends via `-b all`.

Reply:
Neither torch nor array-api-strict are going to work here. This PR is specifically about delegating to "accelerators" , which have a `.scipy` namespace. This rules out torch, and  leaves us with cupyx and jax.numpy.
I test "locally" with jax[CUDA] and CI does jax[cpu].

Reply:
Can we not test that they skip properly or whatever? I should be able to use the `all` backend locally on all of SciPy without issues.

Reply:
```
$ python dev.py test -t scipy/signal/tests/test_signaltools.py::TestConvolve -b array_api_strict -- --tb=no
💻  ninja -C /home/ev-br/repos/scipy/build -j32
ninja: Entering directory `/home/ev-br/repos/scipy/build'
[2/2] Generating scipy/generate-version with a custom command
Build OK
💻  meson install -C build --only-changed --tags runtime,python-runtime,tests,devel
Installing, see meson-install.log...
Installation OK
SciPy from development installed path at: /home/ev-br/repos/scipy/build-install/lib/python3.11/site-packages
Running tests for scipy version:1.14.0.dev0+1320.d555871, installed at:/home/ev-br/repos/scipy/build-install/lib/python3.11/site-packages/scipy
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.11.8, pytest-8.1.1, pluggy-1.4.0
rootdir: /home/ev-br/repos/scipy
configfile: pytest.ini
plugins: anyio-4.3.0, cov-4.1.0, timeout-2.2.0, hypothesis-6.99.2, xdist-3.5.0
collected 16 items                                                                                                                                                                                        

scipy/signal/tests/test_signaltools.py FFFF.F.FF.FFFFF.                                                                                                                                             [100%]

========================================================================================= short test summary info =========================================================================================
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_basic[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_same[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_same_eq[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_complex[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_broadcastable[array_api_strict] - AttributeError: 'Array' object has no attribute 'reshape'
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_2d_arrays[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_input_swapping[array_api_strict] - AttributeError: 'Array' object has no attribute 'reshape'
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_valid_mode2[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_same_mode[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_invalid_shapes[array_api_strict] - AttributeError: 'Array' object has no attribute 'reshape'
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_convolve_method[array_api_strict] - TypeError: Python complex scalars can only be promoted with complex floating-point arrays.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve::test_convolve_method_large_input[array_api_strict] - AssertionError: Namespaces do not match.
====================================================================================== 12 failed, 4 passed in 0.23s =======================================================================================
(scipy-dev) ev-br@qgpu3:~/repos/scipy$ python dev.py test -t scipy/signal/tests/test_signaltools.py::TestConvolve2d -b array_api_strict -- --tb=no
💻  ninja -C /home/ev-br/repos/scipy/build -j32
ninja: Entering directory `/home/ev-br/repos/scipy/build'
[2/2] Generating scipy/generate-version with a custom command
Build OK
💻  meson install -C build --only-changed --tags runtime,python-runtime,tests,devel
Installing, see meson-install.log...
Installation OK
SciPy from development installed path at: /home/ev-br/repos/scipy/build-install/lib/python3.11/site-packages
Running tests for scipy version:1.14.0.dev0+1320.d555871, installed at:/home/ev-br/repos/scipy/build-install/lib/python3.11/site-packages/scipy
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.11.8, pytest-8.1.1, pluggy-1.4.0
rootdir: /home/ev-br/repos/scipy
configfile: pytest.ini
plugins: anyio-4.3.0, cov-4.1.0, timeout-2.2.0, hypothesis-6.99.2, xdist-3.5.0
collected 20 items / 1 deselected / 19 selected                                                                                                                                                           

scipy/signal/tests/test_signaltools.py FFFF..FFFxFxxxFFFF.                                                                                                                                          [100%]

========================================================================================= short test summary info =========================================================================================
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_2d_arrays[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_valid_mode[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_valid_mode_complx[array_api_strict] - ValueError: dtype must be one of the supported dtypes
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_fillvalue[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_wrap_boundary[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_sym_boundary[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_same_with_boundary[symm-expected0-convolve2d-array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_same_with_boundary[wrap-expected1-convolve2d-array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_invalid_shapes[array_api_strict] - AttributeError: 'Array' object has no attribute 'reshape'
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_same_mode[array_api_strict] - AttributeError: module 'array_api_strict' has no attribute 'array'
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_valid_mode2[array_api_strict] - AssertionError: Namespaces do not match.
FAILED scipy/signal/tests/test_signaltools.py::TestConvolve2d::test_consistency_convolve_funcs[array_api_strict] - AttributeError: module 'array_api_strict' has no attribute 'array'
========================================================================== 12 failed, 3 passed, 1 deselected, 4 xfailed in 0.23s ==========================================================================
(scipy-dev) ev-br@qgpu3:~/repos/scipy$ 
```

Can add this many skips of course, if that's really what the ask is.

Reply:
Ah sorry, how Matt did it in `special` was to add a new test file for `_support_alternative_backends` - you may want to follow that. I don't think it makes sense to convert `TestConvolve` yet when the SciPy implementation of convolve is still `np`-only.

Reply:
I'm pretty sure we'd want the skips one way or another, it will become really hard to check on the status of overall backend support if we have to subtract N failures from the overall baseline each time.

Reply:
OK, skips of array_api_strict are in place. 
`$ python dev.py test -t scipy/signal/tests/test_signaltools.py::TestConvolve -v -b all` passes locally.

Now, the whole point of delegating to cupyx/jax here is that the original SciPy test suite passes. I fail to see how a dedicated test file helps in this. In `special`, the relevant test file seems to do something else using hypothesis. Expanding the SciPy test suite of `signaltools` can be a worthy endeavor, but it surely separate.

Reply:
I'll defer to @mdhaber , but one potential viewpoint is that checking the accuracy of each backend's implementation can be deferred at least until we have converted the functions to work with `array_api_strict` - to test the dispatching, we just need something to check that the correct functions are being picked up and arguments are passed correctly.

Reply:
> Now, the whole point of delegating to cupyx/jax here is that the original SciPy test suite passes. I fail to see how a dedicated test file helps in this.

You're right. In `special`, I didn't attempt to ensure that the original SciPy test suite passed with other backends. I just attempted to check that the other backend produces results similar to those of SciPy. It's a smoke test to make sure that we're calling the appropriate function from the other backend and that the signatures are aligned correctly. 

I saw this first... I will look a little closer at what's going on here and come back to this.

Reply:
I think there are two separate things being discussed here:
1. What this dispatch system does with arrays whose libraries don't have a native implementation of the function.
2. How the dispatch system is tested.

I'm not sure that the two are necessarily related.

W.R.T. 1) in `special`, I explicitly converted to `np.ndarray`, executed the function, and converted back to the original array type. Here, the conversion is implicit, and the result is left as a NumPy array if the conversion is needed. IIUC, that's the only reason why the test skips are needed here.

Why not do the conversions? According to ["Support for the array API standard"](https://docs.scipy.org/doc/scipy/dev/api-dev/array_api.html#adding-array-api-standard-support-to-a-scipy-function) document:

![image](https://github.com/scipy/scipy/assets/6570539/f36a2771-ad85-4dc9-a7ae-a4d6a3efa75f)

converting is the preferred approach. It will error out with GPU backends that don't have native implementations, and that's expected. But since both CuPy and JAX have implementations, I don't think you'll need any skips.

Reply:
I see that you already discussed falling back to NumPy here: https://github.com/scipy/scipy/pull/20772#discussion_r1611460831. 

But I think the guidance is to do the conversions - just without explicitly doing a _device transfer_.  And even doing a device transfer might be preferred in some circumstances when `_from_dlpack` makes it into `array-api-compat`. (https://github.com/scipy/scipy/issues/18286#issuecomment-2028756647)

Reply:
> one potential viewpoint is that checking the accuracy of each backend's implementation can be deferred at least until we have converted the functions to work with array_api_strict

FTR, I disagree with this. I believe dispatching to `{backend}.scipy` is only tangentially related to array-api-strict and all that.

> should be able to use the all backend locally on all of SciPy without issues.

OK fine.  `$ python dev.py test -b all` now runs OK on three APIs touched by this PR: `test_signaltools.py::{TestConvolve,TestFFTConvolve,TestConvolve2d`, modulo a couple of small CI error which I'll fix up (curiously, one of them is flagged by smoke-docs :-)). The price is of course the diff size.

> Would property-based tests be feasible for signal?

Don't know TBH. This needs looking into to see if it's possible, and certainly in a separate effort. Changing both tests and implementation is too fiddly.

Reply:
> FTR, I disagree with this. I believe dispatching to {backend}.scipy is only tangentially related to array-api-strict and all that.

perhaps we don't disagree so much - in principal, I agree that it is good to separate them, and in other discussions I have in fact pushed for dispatching work like this to be considered as separate as possible from conversion of our `np`-API-using codebase to the standard API.

My point was a practical one. Right now, the way we test alternative backends, even for dispatching like this, is with the `array_api_compatible` decorator. So at least with our current testing infrastructure, they are very much intertwined, and doing the array-api-strict work before dispatching like this will probably result in less mess skips-wise.

If it makes sense to continue working on dispatching _without_ array-api-strict support in places, while converting the existing tests (rather than new tests like Matt's approach), it may make sense to refactor the way we can test with backends to which we can dispatch. I would be interested if you have any ideas.

Reply:
I'm all for a better term, what would be the one?

On the testing topic, I don't have big asks; the practical point is taken and agreed with. One minor point which  is a bit odd and could maybe be improved, is that multiple `@skip_xp_backends` decorators do not seem to work. Being spoiled by `@pytest.mark.parametrize`, I was naturally expecting to be able to 

```
@skip_xp_backends("torch", reason="dtypes")
@skip_xp_backend("cupy", reason="1D only")
@array_api_compatible
def test_function(self, xp):
     ....
```  

Having to lump the two skips together is a bit awkward (the list of backends is unpacked but the list of `reasons` is not, for one).

Reply:
> multiple @skip_xp_backends decorators do not seem to work

Indeed, this is something I thought about for a while and failed to figure out. It seems a lot of ingenuity is needed to get complex custom pytest stuff like this right. I imagine the decorators will go through a few more iterations - while my changes were an improvement in many ways, there are still a few QoL things like this that aren't ideal or intuitive.

Reply:
> the list of backends is unpacked but the list of reasons is not, for one

if you know a way to unpack arguments behind a kwarg, do share! I suppose an alternative is to take arbitrarily many dicts of `{'backend': 'reason'}` positionally. My refactor didn't do this as the initial version didn't have skip reasons built in at all, so it would have been a more drastic change. Maybe it makes sense now.